### PR TITLE
feat: otter sdk training - how to use the otter sdk

### PR DESCRIPTION
**Pull request by sdo-1A**
## Proposed change

How to use the Otter SDK step
The goal is for the user to create a fetch client and use the pet API to display the results in the UI

## Related issues

*- No issue associated -*

### CODE REVIEW

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  imports: [],
+  templateUrl: './app.component.html',
+  styleUrl: './app.component.scss'
+})
+export class AppComponent {
+  /** Title of the application */
+  public title = 'tutorial-app';
+
+  private readonly petStoreApi = inject(PetApi);
+  public pets = signal<Pet[]>([]);
```
Comment (fpaul-1A) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts:17]:
```suggestion
  public readonly pets = signal<Pet[]>([]);
```

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  constructor() {
+    this.setPets();
+  }
+
+  public setPets() {
+    /* Get the first 10 pets whose status is 'available' */
+    this.petStoreApi.findPetsByStatus({status: 'available'}).then((pets) => {
+      this.pets.set(pets.slice(0, 10));
+    });
+  }
```
Comment (fpaul-1A) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts:28]:
```suggestion
  public async setPets() {
    /* Get the first 10 pets whose status is 'available' */
    const pets = await this.petStoreApi.findPetsByStatus({status: 'available'});
    this.pets.set(pets.slice(0, 10));
  }
```

File: apps/showcase/project.json [hunk 3]
```diff
@@ -96,6 +96,11 @@
             "glob": "*.json",
             "input": "packages/@o3r-training/training-sdk/dist/structure",
             "output": "/assets/@o3r-training/training-sdk/structure"
+          },
+          {
+            "glob": "*.json",
+            "input": "packages/@o3r-training/showcase-sdk/dist/structure",
```
Comment (kpanot) [inline apps/showcase/project.json:102]: not part of this PR touched file but is this path part of the `^structure` inputName files?

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts (hunk 1 above)
Comment (kpanot) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts:17]:
```suggestion
  readonly #pets = signal<Pet[]>([]);
  public readonly pets = this.#pets.asReadonly();
```

(mainly to expose `Signal` and not `writableSignal`, not for the fancy `#` private prefix :))

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  imports: [],
+  templateUrl: './app.component.html',
+  styleUrl: './app.component.scss'
+})
+export class AppComponent {
+  /** Title of the application */
+  public title = 'tutorial-app';
+
+  private readonly petStoreApi = inject(PetApi);
+  public pets = signal<Pet[]>([]);
```
Comment (kpanot) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts:17]: same remark as below

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/instructions.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+#### Using the API object to perform an HTTP request
+Now, you can call the API object to perform the HTTP request you are looking for.\
+As you can see in `app.component.ts`, the `PetApi` has been injected and is ready to be used.\
+For this exercise, you will look for the pets that are of status **available** and display in the UI the names of the first ten pets.
+
+> [!TIP]
+> Have a look at the `sdk/src/api/pet/pet.api.ts` file and look for `findPetsByStatus`.
+
+#### Solution
```
Comment (cpaulve-1A) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/instructions.md:24]: I think the format to show the solution is not the same between this PR and the extended one, let's harmonize it between exercise.

File: apps/showcase/project.json (hunk 3 above)
Comment (sdo-1A) [reply apps/showcase/project.json:102]: https://github.com/AmadeusITGroup/otter/blob/feat/otter-training/nx.json#L95 yes it is

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.html [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<table class="table">
+  <caption align="top">
+    First 10 pets of status <b>available</b> from the Swagger Petstore
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.html:3]:
```suggestion
    First 10 pets with status <b>available</b> from the Swagger Petstore
```

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.html [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    <th scope="col">Name</th>
+    <th scope="col">Status</th>
+  </tr>
+  </thead>
+  <tbody>
+    @for (pet of pets(); track pet.id; let index = $index) {
+      <tr>
+        <td>{{index + 1}}</td>
+        <td><!-- Name of pet --></td>
+        <td><!-- Status of pet --></td>
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.html:17]: I think the exercise should be validated as soon as the developer update the signal with the values retrieved from the API
```suggestion
        <td>{{ pet.name }}</td>
        <td>{{ pet.status }}</td>
```

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.config.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { ApiFetchClient } from '@ama-sdk/client-fetch';
+import { ApplicationConfig, provideZoneChangeDetection, importProvidersFrom } from '@angular/core';
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+import { provideRouter } from '@angular/router';
+import { prefersReducedMotion } from '@o3r/application';
+import { PetApi } from '../../../../libs/sdk/src/api';
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.config.ts:6]: could we avoid doing `../../../..` with a tsconfig path or whatever ?

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  private readonly petStoreApi = inject(PetApi);
+  readonly #pets = signal<Pet[]>([]);
+  public readonly pets = this.#pets.asReadonly();
+
+  constructor() {
+    this.setPets();
+  }
+
+  public setPets() {
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts:24]:
```suggestion
  constructor() {
    void this.setPets();
  }

  public async setPets() {
```

File: packages/@o3r-training/showcase-sdk/package.json [hunk 10]
```diff
@@ -33,13 +33,22 @@
     },
     "./openapi.yml": {
       "default": "./openapi.yml"
+    },
+    "./structure/src.json": {
+      "default": "./structure/src.json"
+    },
+    "./structure/spec.json": {
+      "default": "./structure/src.json"
```
Comment (matthieu-crouzet) [inline packages/@o3r-training/showcase-sdk/package.json:41]:
```suggestion
      "default": "./structure/spec.json"
```

File: apps/showcase/src/assets/trainings/sdk/program.json [hunk 11]
```diff
@@ -7,7 +7,41 @@
...
+        "solutionUrls": [
+          {
+            "path": "./apps/tutorial-app/src/app",
+            "contentUrl": "./steps/typescript-sdk/solution.json"
+          }
+        ],
+        "mode": "interactive",
+        "commands": [
+          "npm install --legacy-peer-deps --ignore-scripts --force",
+          "npm run ng run sdk:build",
```
Comment (fpaul-1A) [inline apps/showcase/src/assets/trainings/sdk/program.json:41]: is that truly necessary?

Comment (sdo-1A) [reply apps/showcase/src/assets/trainings/sdk/program.json]: nope you're right I removed this line

File: packages/@o3r-training/showcase-sdk/project.json [hunk 12]
```diff
@@ -36,6 +36,21 @@
...
         "command": "npm publish packages/@o3r-training/showcase-sdk/dist"
       }
+    },
+    "extract-folder-structure": {
+      "cache": true,
+      "executor": "nx:run-script",
+      "options": {
+        "script": "extract"
+      },
+      "inputs": [
```
Comment (kpanot) [inline packages/@o3r-training/showcase-sdk/project.json:46]: The following file seems missing `{projectRoot}/openapitools.json`
(if not already part of `source`)

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.config.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+const metaReducers = [storageSync.localStorageSync()];
+const runtimeChecks: Partial<RuntimeChecks> = {
+  strictActionImmutability: false,
+  strictActionSerializability: false,
+  strictActionTypeUniqueness: !environment.production,
+  strictActionWithinNgZone: !environment.production,
+  strictStateImmutability: !environment.production,
+  strictStateSerializability: false
+};
```
[the remark this answers is not in the document]
Comment (fpaul-1A) [reply apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.config.ts]: we are already installing using `--legacy-peer-deps` so I don't think the warnings will matter
there is no reason for `@ngrx` to be used in the training so we can safely remove the dependency

**Review by fpaul-1A — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  imports: [],
+  templateUrl: './app.component.html',
+  styleUrl: './app.component.scss'
+})
+export class AppComponent {
+  /** Title of the application */
+  public title = 'tutorial-app';
+
+  private readonly petStoreApi = inject(PetApi);
+  readonly #pets = signal<Pet[]>([]);
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts:16]:
```suggestion
  public readonly #pets = signal<Pet[]>([]);
```

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  imports: [],
+  templateUrl: './app.component.html',
+  styleUrl: './app.component.scss'
+})
+export class AppComponent {
+  /** Title of the application */
+  public title = 'tutorial-app';
+
+  private readonly petStoreApi = inject(PetApi);
+  readonly #pets = signal<Pet[]>([]);
```
Comment (matthieu-crouzet) [inline apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/solution/app.component.ts:16]:
```suggestion
  private readonly #pets = signal<Pet[]>([]);
```

File: apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts (hunk 14 above)
Comment (sdo-1A) [reply apps/showcase/src/assets/trainings/sdk/steps/typescript-sdk/exercise/app.component.ts:16]: the `#` prefix means that the variable is private, so there is no need to rewrite `private` before (in fact it causes an error)

**Review by matthieu-crouzet — APPROVED** [review]
